### ISSUE TITLE
TypeInstantiated event not raised when registering using createServiceFunc

### ISSUE THREAD

**Issue by ttustonic** [Catel/Catel#1346]
### Feature description

Currently, when registering type using `createServiceFunc `parameter, the `TypeInstantiated `event is not raised. Perhaps there should be a wrapper around this function (as well as the default one) which would raise this event ?

**Comment by GeertvanHorrik** [comment]
Good point. But the createServiceFunc is purely for instantiating the type, but what if the event gets raised if the function returns a non-null result?

**Comment by ttustonic** [comment]
Hm, I don't quite understand. Here's what I had in mind:

in `private void RegisterType`:

```
registeredTypeInfo = 
 new ServiceLocatorRegistration(serviceType, serviceImplementationType, tag, registrationType,
    x => CreateServiceInstanceWrapper(createServiceFunc ?? DefaultCreateServiceFunc, x));
```

and then:

```
private object DefaultCreateServiceFunc(ServiceLocatorRegistration registration)
{
        var instance = _typeFactory.CreateInstanceWithTag(registration.ImplementingType, registration.Tag);
        return instance;
}

private object CreateServiceInstanceWrapper(Func<ServiceLocatorRegistration, object>, 
createServiceFunc, ServiceLocatorRegistration registration)
{
        var instance = createServiceFunc(registration);
        if (instance is null)
        {
             ThrowTypeNotRegisteredException(registration.DeclaringType, "Failed to instantiate the type using the TypeFactory. Check if the required dependencies are registered as well or that the type has a valid constructor that can be used.");
        }
        var handler = TypeInstantiated;
        if (handler != null)
        {
            handler(this, new TypeInstantiatedEventArgs(registration.DeclaringType, 
           registration.ImplementingType, 
           registration.Tag, registration.RegistrationType, instance));
            }
            return instance;
        }
```

So it works the same for the default case, and if there's a `createServiceFunc`, it just wraps it and raises the event, just like in the default.

**Comment by GeertvanHorrik** [comment]
Good point, sounds like a plan!

**Comment by GeertvanHorrik** [comment]
Perfect, thanks! Will be available in Catel Build - 5.11.0-alpha.102
